File: src/registry/registry-session.ts
import { hostsFingerprint, type AppConfig, type HostEntry } from "../config/host-config.js";
import { errorText } from "../error-text.js";
import type { RegistrySnapshot } from "./paseo-registry.js";

export interface RegistrySession {
  /** Reads once and applies. Never rejects. */
  start(): Promise<void>;
  /** Re-reads. Never rejects. Exposed for the watcher, the poll, and tests. */
  refresh(): Promise<void>;
  /** The fleet's unusable entries, for the half of the error row it owns. */
  noteEntryFailures(failures: string[]): void;
  stop(): void;
}

const NO_HOSTS_MESSAGE = "No hosts yet. Pair a host in the Paseo app.";

/**
 * Owns the tray's view of the Paseo app's host registry: when to re-read it,
 * whether anything changed, and what the error row says.
 *
 * Electron-free on purpose, the same way `host-fleet.ts` is. Everything here
 * is reachable on a machine where the Paseo app is missing, mid-write, or has
 * moved its storage, so it has to be testable without a main process.
 *
 * Nothing here may reject. A menu-bar app that dies on a torn read of another
 * program's database leaves the user nothing to fix it with.
 */
export function createRegistrySession(options: {
  /** Injected so tests need no filesystem. Production passes `readRegistry`. */
  readRegistry: () => Promise<RegistrySnapshot | null>;
  /** Starts watching; returns the stop function. Production watches the dir. */
  watch: (onChange: () => void) => () => void;
  applyConfig: (config: AppConfig) => Promise<void>;
  onConfigError: (message: string | null) => void;
  /** Safety net for events the watcher misses. Zero disables it. */
  pollMs?: number;
  debounceMs?: number;
}): RegistrySession {
  const { readRegistry, watch, applyConfig, onConfigError } = options;
  const pollMs = options.pollMs ?? 60_000;
  const debounceMs = options.debounceMs ?? 500;

  let appliedFingerprint: string | null = null;
  let stopWatching: (() => void) | null = null;
  let debounceTimer: NodeJS.Timeout | null = null;
  let pollTimer: NodeJS.Timeout | null = null;
  let running: Promise<void> | null = null;

  // Two independent problems, reported through one menu row: reading the
  // registry, and entries the fleet could not use. Either can be fixed without
  // the other, so neither may clear the other.
  let registryError: string | null = null;
  let fleetError: string | null = null;

  function refreshConfigError(): void {
    const problems = [registryError, fleetError].filter((problem) => problem !== null);
    onConfigError(problems.length > 0 ? problems.join("\n\n") : null);
  }

  async function readAndApply(): Promise<void> {
    let snapshot: RegistrySnapshot | null;
    try {
      snapshot = await readRegistry();
    } catch (error) {
      // Keep the last known-good host set live and say what went wrong. A
      // compaction mid-read lands here and resolves itself on the next tick.
      registryError = errorText(error);
      refreshConfigError();
      return;
    }

    const hosts: HostEntry[] = snapshot?.hosts ?? [];
    const failures = snapshot?.failures ?? [];

    const problems: string[] = [];
    if (snapshot === null) problems.push(NO_HOSTS_MESSAGE);
    if (failures.length > 0) {
      problems.push(`These hosts could not be used:\n\n${failures.join("\n")}`);
    }
    registryError = problems.length > 0 ? problems.join("\n\n") : null;
    refreshConfigError();

    // Rebuilding tears down live connections, so it happens only when the
    // host set genuinely differs -- Chromium rewrites this database constantly
    // for keys we do not care about.
    const fingerprint = hostsFingerprint(hosts);
    if (fingerprint === appliedFingerprint) return;
    appliedFingerprint = fingerprint;
    await applyConfig({ version: 1, hosts });
  }

  /** Serializes reads so a watcher burst cannot interleave two applies. */
  function serialize(): Promise<void> {
    const next = (running ?? Promise.resolve()).then(readAndApply, readAndApply);
    running = next.catch(() => undefined);
    return next;
  }

  async function safeRefresh(): Promise<void> {
    try {
      await serialize();
    } catch (error) {
      // serialize() already routes read failures to the error row; this is the
      // last line against applyConfig throwing.
      registryError = errorText(error);
      refreshConfigError();
    }
  }

  return {
    async start() {
      stopWatching = watch(() => {
        if (debounceTimer) clearTimeout(debounceTimer);
        debounceTimer = setTimeout(() => {
          debounceTimer = null;
          void safeRefresh();
        }, debounceMs);
      });
      if (pollMs > 0) {
        pollTimer = setInterval(() => void safeRefresh(), pollMs);
        // A background poll must never hold the process open on its own.
        pollTimer.unref?.();
      }
      await safeRefresh();
    },

    refresh() {
      return safeRefresh();
    },

    noteEntryFailures(failures) {
      fleetError =
        failures.length > 0 ? `These hosts could not be used:\n\n${failures.join("\n")}` : null;
      refreshConfigError();
    },

    stop() {
      if (debounceTimer) clearTimeout(debounceTimer);
      if (pollTimer) clearInterval(pollTimer);
      debounceTimer = null;
      pollTimer = null;
      stopWatching?.();
      stopWatching = null;
    },
  };
}

File: PaseoIconPackage/Sources/PaseoIconCore/Registry/RegistrySession.swift
import Foundation

/// Owns the tray's view of the Paseo app's host registry: when to re-read
/// it, whether anything changed, and what the error row says. Nothing here
/// throws out of `start` or `refresh`: a menu bar app that dies on a torn
/// read of another program's database leaves the user nothing to fix it
/// with.
@MainActor
public final class RegistrySession {
    public static let noHostsMessage = "No hosts yet. Pair a host in the Paseo app."

    private let readRegistry: () async throws -> RegistrySnapshot?
    private let watch: (@escaping () -> Void) -> () -> Void
    private let applyConfig: (AppConfig) async throws -> Void
    private let onConfigError: (String?) -> Void
    private let afterRead: (() -> Void)?
    private let pollInterval: Duration
    private let debounce: Duration
    private let clock: any Clock<Duration>

    private var appliedFingerprint: String?
    private var stopWatching: (() -> Void)?
    private var debounceTask: Task<Void, Never>?
    private var pollTask: Task<Void, Never>?
    private var chain: Task<Void, Never>?

    // Two independent problems, reported through one menu row: reading the
    // registry, and entries the fleet could not use. Neither may clear the other.
    private var registryError: String?
    private var fleetError: String?

    /// - Parameters:
    ///   - readRegistry: production passes `PaseoRegistry.read`.
    ///   - watch: starts watching; returns the stop function.
    ///   - afterRead: runs after every read, failed ones included. Production
    ///     re-attaches the directory watch here, which is what makes installing
    ///     Paseo mid-session take effect on the next poll.
    ///   - pollInterval: safety net for events the watcher misses. Zero disables it.
    public init(
        readRegistry: @escaping () async throws -> RegistrySnapshot?,
        watch: @escaping (@escaping () -> Void) -> () -> Void,
        applyConfig: @escaping (AppConfig) async throws -> Void,
        onConfigError: @escaping (String?) -> Void,
        afterRead: (() -> Void)? = nil,
        pollInterval: Duration = .seconds(60),
        debounce: Duration = .milliseconds(500),
        clock: any Clock<Duration> = ContinuousClock()
    ) {
        self.readRegistry = readRegistry
        self.watch = watch
        self.applyConfig = applyConfig
        self.onConfigError = onConfigError
        self.afterRead = afterRead
        self.pollInterval = pollInterval
        self.debounce = debounce
        self.clock = clock
    }

    /// Reads once and applies. Never throws.
    public func start() async {
        stopWatching = watch { [weak self] in self?.scheduleDebouncedRefresh() }
        if pollInterval > .zero {
            pollTask = Task { [weak self] in
                while true {
                    guard let self else { return }
                    do { try await self.clock.sleep(for: self.pollInterval) } catch { return }
                    await self.refresh()
                }
            }
        }
        await refresh()
    }

    /// Re-reads. Never throws. Reads are serialized so a watcher burst cannot
    /// interleave two applies.
    public func refresh() async {
        let previous = chain
        let task = Task { [weak self] in
            await previous?.value
            await self?.readAndApply()
        }
        chain = task
        await task.value
    }

    /// The fleet's unusable entries, for the half of the error row it owns.
    public func noteEntryFailures(_ failures: [String]) {
        fleetError = failures.isEmpty ? nil : Self.describeUnusableHosts(failures)
        refreshConfigError()
    }

    public func stop() {
        debounceTask?.cancel()
        debounceTask = nil
        pollTask?.cancel()
        pollTask = nil
        // The read chain too. On quit `AppCoordinator.stop()` closes the fleet
        // while a read may still be awaiting its detached LevelDB work; when it
        // resumed it called `applyConfig` and rebuilt every connection during
        // teardown. The process is going away either way, but shutdown should
        // mean what it says.
        chain?.cancel()
        chain = nil
        stopWatching?()
        stopWatching = nil
    }

    private func scheduleDebouncedRefresh() {
        debounceTask?.cancel()
        debounceTask = Task { [weak self] in
            guard let self else { return }
            do { try await self.clock.sleep(for: self.debounce) } catch { return }
            self.debounceTask = nil
            await self.refresh()
        }
    }

    private static func describeUnusableHosts(_ failures: [String]) -> String {
        "These hosts could not be used:\n\n\(failures.joined(separator: "\n"))"
    }

    private func refreshConfigError() {
        let problems = [registryError, fleetError].compactMap { $0 }
        onConfigError(problems.isEmpty ? nil : problems.joined(separator: "\n\n"))
    }

    /// One read, start to finish, that cannot throw.
    private func readAndApply() async {
        await readOnce()
        afterRead?()
    }

    private func readOnce() async {
        let snapshot: RegistrySnapshot?
        do {
            snapshot = try await readRegistry()
        } catch {
            // Keep the last known-good host set live and say what went wrong.
            registryError = errorText(error)
            refreshConfigError()
            return
        }

        let hosts = snapshot?.hosts ?? []
        let failures = snapshot?.failures ?? []

        var problems: [String] = []
        // Absent key and empty array are the same dead end for the user.
        if hosts.isEmpty && failures.isEmpty { problems.append(Self.noHostsMessage) }
        if let warning = snapshot?.warning { problems.append(warning) }
        if !failures.isEmpty { problems.append(Self.describeUnusableHosts(failures)) }

        // The host set is hand-built from another program's storage, so it is
        // validated before the fleet sees it; this is the only path that builds a config.
        let config: AppConfig
        do {
            config = try AppConfig.validate(hosts: hosts)
        } catch {
            problems.append("The Paseo app's host list could not be used:\n\n\(errorText(error))")
            registryError = problems.joined(separator: "\n\n")
            refreshConfigError()
            return
        }

        registryError = problems.isEmpty ? nil : problems.joined(separator: "\n\n")
        refreshConfigError()

        // Rebuilding tears down live connections, so only when the host set differs.
        let fingerprint = hostsFingerprint(config.hosts)
        if fingerprint == appliedFingerprint { return }
        do {
            try await applyConfig(config)
            // Recorded only once the fleet has actually taken it.
            appliedFingerprint = fingerprint
        } catch {
            // The fleet may be half torn down, so nothing counts as applied any more.
            appliedFingerprint = nil
            problems.append("The hosts could not be applied:\n\n\(errorText(error))")
            registryError = problems.joined(separator: "\n\n")
            refreshConfigError()
        }
    }
}
